htlcinterceptor/middleware.go

```diff
@@ -0,0 +1,147 @@
+package htlcinterceptor
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	//ErrFwdNotExists is an error returned when the caller tries to resolve
+	// a forward that doesn't exist anymore.
+	ErrFwdNotExists = errors.New("forward does not exist")
+)
+
+// Middleware is a middleware for other sub systems, including the RPC layer,
+// to be able to intercept htlc forwards. It holds a list of handlers that are
+// invoked sequentially per forward to figure out if one of them is interested
+// to handle the forward outside the switch. In addition it exposed API for
+// resolving existing hold forward.
+type Middleware struct {
+	sync.RWMutex
+	// ForwardHtlcHandler is the callback that is called for each forward of
+	// incoming htlc. It should return true if it is interested in handling
+	// the forwardHtlc.
+	fwdHandlers  map[uint64]ForwardHtlcHandler
+	fwdHandlerID uint64
+
+	// holdForwards is a map of current hold forwards and their corresponding
+	// ForwardResolver.
+	holdForwards map[channeldb.CircuitKey]ForwardResolver
+}
+
+// NewMiddleware initializes a Middleware.
+func NewMiddleware() *Middleware {
+	return &Middleware{
+		fwdHandlers:  make(map[uint64]ForwardHtlcHandler),
+		holdForwards: make(map[channeldb.CircuitKey]ForwardResolver),
+	}
+}
+
+// AddForwardHtlcHandler adds a ForwardHtlcHandler.
+func (c *Middleware) AddForwardHtlcHandler(
+	handler ForwardHtlcHandler) uint64 {
+
+	c.Lock()
+	defer c.Unlock()
+	c.fwdHandlerID++
+	c.fwdHandlers[c.fwdHandlerID] = handler
+
+	// Return the id so that a caller can call RemoveForwardHtlcHandler.
+	return c.fwdHandlerID
+}
+
+// RemoveForwardHtlcHandler removes a ForwardHtlcHandler.
+func (c *Middleware) RemoveForwardHtlcHandler(id uint64) {
+	c.Lock()
+	delete(c.fwdHandlers, id)
+	c.Unlock()
+}
+
+// InterceptForwardHtlc is called before each forward and query its registered
+// handlers to see if one of them is interested in handling (hold) this
+// forward. In case one of them returns true the forward will be marked as hold
+// and signals the switch to wait for a later resolve action which should be
+// done later by calling one of:
+// * ResumeHoldForward
+// * SettleHoldForward
+// * FailHoldForward.
+func (c *Middleware) InterceptForwardHtlc(
+	circuitKey channeldb.CircuitKey,
+	htlc lnwire.UpdateAddHTLC, resolver ForwardResolver) bool {
+
+	c.Lock()
+	defer c.Unlock()
+	for _, actionResolver := range c.fwdHandlers {
```

Comment:
Are multiple handlers even needed at all for the described use cases? Without it, the `Middleware` may not be needed at all, simplifying the first version of this

Reply:
@joostjager I think this is a valid point. This can be simplified to support only one interceptor. In case an RPC interceptor is already registered, a new RPC one would probably need to disconnect the previous stream or reject the new registration.
@Roasbeef what do you think? I prefer we all agree on this before going for it.

Reply:
I will simplify it now. We can only support more interceptors later.

Reply:
I have pushed a new version that support only one interceptor at a time.
This simplifies the implementation as I removed the middleware and replaced it by a thin implementation at the RPC layer. 
One more change was to use the same bidirectional stream for the client to resolve packets.

So in general the flows goes like this:
1. Link forward a packet, packet intercepted and sent to the RPC layer.
2. For each packet the RPC layer do the following:

- Hold the packet immediately and signal hold backwards to the switch.
- Send the packet for the client

3. In parallel RPC listens to client resolutions responses on the same stream and resolve packets accordingly.
4. On disconnection all remaining held packets are resolved using the default behavior (currently `Resume`)